src/types.ts

```diff
@@ -1,19 +1,29 @@
-export type Result<T> = { type: "error"; value: string } | { type: "success"; value: T };
+export type Result<E, T> = { type: "error"; value: E } | { type: "success"; value: T };
 
 /**
  * Successful result constructor
  */
-export const success = <T>(value: T): Result<T> => ({
+export const success = <E, T>(value: T): Result<E, T> => ({
   type: "success",
   value,
 });
 
 /**
  * Error result constructor
  */
-export const error = <T>(value: string): Result<T> => ({
+export const error = <E, T>(value: E): Result<E, T> => ({
   type: "error",
   value,
 });
 
-export type Decoder<T> = (a: any) => Result<T>;
+/**
+ * Decoder error
+ */
+export interface DecodeError {
+  path: string[];
+  expectType: string;
+  received: any;
+  error?: string;
+}
```

Comment:
I think `expectType` is useful for most of the cases - all I was saying was that if you use `ensure`, you will get `expectType: "ensured"` instead of an original string, which still holds some information. Then you can look at the `error` field and see what it was exactly.

I see other libraries do things like `expected string, received: ""`, which I wouldn't want to do for primitive types. `expectType` and `received` would contain the same information, just without the need for parsing.

Proposal_2 is great, I'll add that right away.

Proposal_3 is also great, though it should go in a different PR. Would you mind outlining it in an issue?

Reply:
regarding proposal_1
did not get the counter-arguments, could you pls rephrase in bullet points or smth, thanks!

Reply:
You never do - makes sense, good simplification :)